agent/artifact_uploader.go

```diff
@@ -125,19 +125,26 @@ func (a *ArtifactUploader) Collect(ctx context.Context) (artifacts []*api.Artifa
 
 		a.logger.Debug("Searching for %s", globPath)
 
-		// Resolve the globs (with * and ** in them), if it's a non-globbed path and doesn't exists
-		// then we will get the ErrNotExist that is handled below
-		globfunc := zglob.Glob
-		if a.conf.GlobResolveFollowSymlinks {
-			// Follow symbolic links for files & directories while expanding globs
-			globfunc = zglob.GlobFollowSymlinks
+		// Resolve the globs (with * and ** in them)
+		pattern, err := zzglob.Parse(globPath)
+		if err != nil {
+			return nil, fmt.Errorf("invalid glob pattern: %w", err)
 		}
-		files, err := globfunc(globPath)
-		if errors.Is(err, os.ErrNotExist) {
-			a.logger.Info("File not found: %s", globPath)
-			continue
-		} else if err != nil {
-			return nil, fmt.Errorf("resolving glob: %w", err)
+		var files []string
+		err = pattern.Glob(func(path string, d fs.DirEntry, err error) error {
+			if err != nil {
+				a.logger.Warn("Couldn't walk path %s", path)
+				return nil
+			}
+			if d != nil && d.IsDir() {
+				a.logger.Warn("Glob pattern %s matched a directory %s", globPath, path)
+				return nil
+			}
+			files = append(files, path)
+			return nil
+		}, a.conf.GlobResolveFollowSymlinks)
+		if err != nil {
+			return nil, fmt.Errorf("globbing pattern: %w", err)
```

Comment:
I'm quite confident that this glob will be better, but I think the safe approach is to put it behind an experiment.